flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/remote/RemoteExecutor.java

```diff
@@ -0,0 +1,423 @@
+                            "Could not find a result with result identifier '%s'.", resultId));
+        }
+
+        // stop retrieval and remove the result
+        LOG.info("Cancelling job {} and result retrieval.", resultId);
+        try {
+            // this operator will also stop flink job
+            result.close();
+        } catch (Throwable t) {
+            throw new SqlExecutionException("Could not cancel the query execution", t);
+        }
+        resultStore.removeResult(resultId);
+    }
+
+    // ---------------------------------------------------------------------------------------------
+    private <
+                    M extends MessageHeaders<R, P, U>,
+                    U extends MessageParameters,
+                    R extends RequestBody,
+                    P extends ResponseBody>
+            CompletableFuture<P> sendRequest(M messageHeaders, U messageParameters, R request)
+                    throws IOException {
+        return restClient.sendRequest(address, port, messageHeaders, messageParameters, request);
+    }
+
+    private FetchResultsResponseBody fetchResults(OperationHandle operationHandle) {
+        return fetchResults(operationHandle, 0L);
+    }
+
+    public FetchResultsResponseBody fetchResults(OperationHandle operationHandle, long token)
+            throws SqlClientException {
+        FetchResultsTokenParameters fetchResultsTokenParameters =
+                new FetchResultsTokenParameters(sessionHandle, operationHandle, token);
+        try {
+            return sendRequest(
+                            FetchResultsHeaders.getInstance(),
+                            fetchResultsTokenParameters,
+                            EmptyRequestBody.getInstance())
+                    .get();
+        } catch (Exception e) {
+            LOG.error(
+                    String.format(
+                            "Unexpected error occurs when fetching results. OperationHandle ID: %s.",
+                            operationHandle),
+                    e);
+            throw new SqlClientException(
+                    String.format(
+                            "Unexpected error occurs when fetching results. OperationHandle ID: %s.",
+                            operationHandle),
+                    e);
+        }
+    }
+
+    @SuppressWarnings("BusyWait")
+    private FetchResultsResponseBody fetchWhenResultsReady(
+            OperationHandle operationHandle, Duration timeout) throws SqlClientException {
+        long timeoutMs = timeout.toMillis();
+        long startingTime = System.currentTimeMillis();
+
+        Function<FetchResultsResponseBody, Boolean> wait =
+                response -> response.getResultType().equals(ResultSet.ResultType.NOT_READY.name());
+        FetchResultsResponseBody response = fetchResults(operationHandle);
+
+        while (wait.apply(response) && System.currentTimeMillis() - startingTime < timeoutMs) {
+            try {
+                Thread.sleep(10L);
```

Comment:
From local testing with the POC branch, I found the 10 millisecond polling a bit aggressive.  

Like the other comment in this PR, I'd suggest externalize the polling periods as a configuration.

I think the heartbeat polling should be separate from one for getting command results.  In my opinions, heart beats could be quite slow (several seconds).  

For commands/results, longer delays would definitely slow down the client.  On the other hand, too high and unnecessary network traffic would be generated.